### PR TITLE
NOT and IN filter

### PR DESCRIPTION
**Pull request by stapel**
This would allow you to use the NOT- and IN-operators, in SqlWhereBuilder you could write: ``!(Person::name `in` listOf("Peter", "Paul"))``

### CODE REVIEW

**Comment by stapel** [comment]
I am not so sure about the `Collection`-type in InFilter though.

**Comment by stapel** [comment]
Also, `foo in ()` (empty list) does not seem to be valid SQL, but I don't know what to do there.

**Comment by mvysny** [comment]
Hey, thanks for the PR, looks great! Let me review.

> foo in ()

This is basically equal to `false` since it does not match anything. I wonder whether we should ban such a filter... I mean, we can let concrete data loader implementors such as vok-orm deal with producing valid SQLs... I'd propose to simply ban empty collections in the `InFilter` class, and maybe revert this decision in the future when a compelling use-case is found.

**Review by mvysny — CHANGES_REQUESTED** [review]
Looks great! I have added a couple of items to discuss, could you please look into those?

File: src/main/kotlin/com/github/mvysny/vokdataloader/Filters.kt [hunk 1]
```diff
@@ -218,6 +233,8 @@ class SqlWhereBuilder<T: Any>(val clazz: Class<T>) {
     @Suppress("UNCHECKED_CAST")
     infix fun <R> KProperty1<T, R?>.gt(value: R): Filter<T> =
         OpFilter(name, value as Comparable<Any>, CompareOperator.gt)
+    @Suppress("UNCHECKED_CAST")
+    infix fun <R> KProperty1<T, R>.`in`(value: R): Filter<T> = InFilter(name, value as Collection<Comparable<*>>)
```
Comment (mvysny) [inline src/main/kotlin/com/github/mvysny/vokdataloader/Filters.kt:281]: Maybe the `in` function could only accept `Collection<R>` as a `value`?

File: src/main/kotlin/com/github/mvysny/vokdataloader/Filters.kt [hunk 2]
```diff
@@ -75,6 +76,15 @@ data class OpFilter<T: Any>(override val propertyName: DataLoaderPropertyName, o
     override fun test(t: T): Boolean = operator.test(getValue(t) as Comparable<Any>?, value)
 }
 
+data class InFilter<T: Any>(
```
Comment (mvysny) [inline src/main/kotlin/com/github/mvysny/vokdataloader/Filters.kt:86]: Could you please add kdoc? Something along these lines:

> only matches items present in the [value] collection. For best in-memory performance we encourage you to only pass in [Set]. The in-memory matching is performed based on Collection.contains; other data providers may specify different criteria for comparison.

File: src/main/kotlin/com/github/mvysny/vokdataloader/Filters.kt [hunk 3]
```diff
@@ -156,6 +166,11 @@ class ILikeFilter<T: Any>(override val propertyName: DataLoaderPropertyName, sta
     }
 }
 
+class NotFilter<T: Any>(val child: Filter<in T>) : Filter<T> {
```
Comment (mvysny) [inline src/main/kotlin/com/github/mvysny/vokdataloader/Filters.kt:191]: Could you please add a very simple kdoc, e.g. "Matches anything that is not matched by the [child] filter. May not be supported by all data loaders, e.g. REST endpoints typically don't allow for this kind of filters"

**Review by mvysny — COMMENTED** [review]
Awesome, thank you!

**Review by mvysny — APPROVED** [review]

**Comment by stapel** [comment]
Thanks, I had some troubles seeing your comments, I only got them via mail at first.
Also thanks for all your work!

**Comment by mvysny** [comment]
Thank you for your contribution :+1:  I'm also new to this Github review system, it looked chatty and messy to me as well, I hope if I did that correctly :)
Please let me know if this functionality is critical for you - if yes, I'll release a new version.

**Pull request merged**